crypto/chacha_drbg.c

```diff
@@ -19,44 +19,107 @@
 
 #include "chacha_drbg.h"
 
+#include <assert.h>
 #include <stdint.h>
+#include <string.h>
+
+#include "chacha20poly1305/ecrypt-portable.h"
+#include "memzero.h"
+#include "sha2.h"
+
+#define CHACHA_DRBG_KEY_LENGTH 32
+#define CHACHA_DRBG_COUNTER_LENGTH 8
+#define CHACHA_DRBG_IV_LENGTH 8
+#define CHACHA_DRBG_SEED_LENGTH \
+  (CHACHA_DRBG_KEY_LENGTH + CHACHA_DRBG_COUNTER_LENGTH + CHACHA_DRBG_IV_LENGTH)
 
 #define MAX(a, b) (a) > (b) ? (a) : (b)
 
-void chacha_drbg_init(CHACHA_DRBG_CTX *ctx,
-                      const uint8_t entropy[CHACHA_DRBG_SEED_LENGTH]) {
+static void derivation_function(const uint8_t *input1, size_t input1_length,
+                                const uint8_t *input2, size_t input2_length,
+                                uint8_t *output, size_t output_length) {
+  // Implementation of Hash_df from NIST SP 800-90A
+  uint32_t block_count = (output_length - 1) / SHA256_DIGEST_LENGTH + 1;
+  assert(block_count <= 255);
+
+  SHA256_CTX ctx = {0};
+
+  for (uint8_t counter = 1; counter <= block_count; counter++) {
+    uint32_t output_length_bits = output_length * 8;
+#if BYTE_ORDER == LITTLE_ENDIAN
+    REVERSE32(output_length_bits, output_length_bits);
+#endif
+
+    sha256_Init(&ctx);
+    sha256_Update(&ctx, &counter, sizeof(counter));
+    sha256_Update(&ctx, (uint8_t *)&output_length_bits,
+                  sizeof(output_length_bits));
+    sha256_Update(&ctx, input1, input1_length);
+    sha256_Update(&ctx, input2, input2_length);
+
+    if (counter != block_count) {
+      sha256_Final(&ctx, output);
+      output += SHA256_DIGEST_LENGTH;
+    } else {  // the last block
+      uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+      sha256_Final(&ctx, digest);
+      memcpy(output, digest, output_length % SHA256_DIGEST_LENGTH);
```

Comment:
Looks like a bug: If `output_length` is a multiple of `SHA256_DIGEST_LENGTH`, then we lose the last block of output, don't we?